### ISSUE TITLE
Cloudflare CDN offers expired cert from CN = ssl392509.cloudflaressl.com

### ISSUE THREAD

**Issue by agilmore2** [Cisco-Talos/clamav#555]
I see SSL expired certificates for 104.16.219.84/db.us.clamav.net/database.clamav.net 

Similar report in https://forum.efa-project.org/viewtopic.php?t=4491#p17028

This breaks freshclam downloads as my layer 7 proxies refuse to support expired certificates.

I get output like
```
depth=2 C = GB, ST = Greater Manchester, L = Salford, O = COMODO CA Limited, CN = COMODO ECC Certification Authority
verify return:1
depth=1 C = GB, ST = Greater Manchester, L = Salford, O = COMODO CA Limited, CN = COMODO ECC Domain Validation Secure Server CA 2
verify return:1
depth=0 CN = ssl392509.cloudflaressl.com
verify error:num=10:certificate has expired
notAfter=Oct 13 23:59:59 2020 GMT
verify return:1
depth=0 CN = ssl392509.cloudflaressl.com
notAfter=Oct 13 23:59:59 2020 GMT
verify return:1

```
from openssl s_client -connect 104.16.219.84.net:443 -proxy <proxy> -showcerts | openssl x509 -text -noout

Source IP is located in the USA.

**Comment by agilmore2** [comment]
Also from 104.16.218.84 as a server

**Comment by micahsnyder** [comment]
We discussed this a bit internally. It seems like a combination of issues.  

1. It seems like your environment's proxy isn't doing SNI routing.  It isn't providing the server name when checking the certificate for the IP.  Cloudflare then doesn't know to send back the certificate for "database.clamav.net" so it hands back a default certificate.  That leads into ....

2. Cloudflare's default certificate appears to be invalid/expired. We have opened an issue with Cloudflare to ask about this.

So if you do this, you get the default cert and it is expired:

```
❯ openssl s_client -connect 104.16.218.84:443  -showcerts | openssl x509 -text -noout
CONNECTED(00000003)
Can't use SSL_get_servername
depth=2 C = GB, ST = Greater Manchester, L = Salford, O = COMODO CA Limited, CN = COMODO ECC Certification Authority
verify return:1
depth=1 C = GB, ST = Greater Manchester, L = Salford, O = COMODO CA Limited, CN = COMODO ECC Domain Validation Secure Server CA 2
verify return:1
depth=0 CN = ssl392509.cloudflaressl.com
verify error:num=10:certificate has expired
notAfter=Oct 13 23:59:59 2020 GMT
verify return:1
depth=0 CN = ssl392509.cloudflaressl.com
notAfter=Oct 13 23:59:59 2020 GMT
verify return:1
....
continues on to show the default cert (doesn't have a X509v3 Subject Alternative Name)
```

And this, the cert is good:
```
❯ openssl s_client -connect 104.16.218.84:443  -servername database.clamav.net   -showcerts | openssl x509 -text -noout
CONNECTED(00000003)
depth=2 C = IE, O = Baltimore, OU = CyberTrust, CN = Baltimore CyberTrust Root
verify return:1
depth=1 C = US, O = "Cloudflare, Inc.", CN = Cloudflare Inc ECC CA-3
verify return:1
depth=0 C = US, ST = California, L = San Francisco, O = "Cloudflare, Inc.", CN = sni.cloudflaressl.com
verify return:1
....
continues on to show the cert for database.clamav.net   (see the X509v3 Subject Alternative Name)
```

So to sum up, I think a configuration change for your proxy would fix the issue on your end, but we'll also see what comes of the support ticket with Cloudflare.  Maybe they can fix the default cert.

**Comment by micahsnyder** [comment]
We heard back from Cloudflare. The response is essentially that the expired certificate indicates to them that you're not specifying the servername. They wrote:

> if you're not specifying the SAN correctly then we don't consider this a bug

They didn't say if the expired cert is intentional, but it almost sounds like it is, so they can quickly triage the SAN / SNI configuration issue.

I saw in Discord chat that you've made adjustments to your proxy and are now past this specific issue. So I think we can close this.
